# HG changeset patch
# User Markus Mützel <markus.muetzel@gmx.de>
# Date 1774544797 -3600
#      Thu Mar 26 18:06:37 2026 +0100
# Node ID d375dcc90801bc1d2c4d54e9bdc80a0df097afdd
# Parent  b7bade3864deecd376187d99725c51aed0270352
Use const qualifiers for pointers to arrays with indices.

* src/sparsersb.cc (rsboi_mtx_alloc_from_csc_const,
rsboi_mtx_alloc_from_coo_const): Use const qualifiers for read-only arguments.
Use local buffer instead of overwriting the index array in Octave values.
This avoids a compilation error with Octave 12. See hg id 9733eeb03fda of the
Octave repository.

diff --git a/src/sparsersb.cc b/src/sparsersb.cc
--- a/src/sparsersb.cc
+++ b/src/sparsersb.cc
@@ -326,10 +326,9 @@
 	return errval;
 }
 
-static void rsboi_oi2ri( octave_idx_type * IP, rsb_nnz_idx_t nnz)
+static void rsboi_oi2ri( const octave_idx_type * IP, rsb_coo_idx_t * RP, rsb_nnz_idx_t nnz)
 {
 	// octave_idx_type -> rsb_coo_idx_t
-	rsb_coo_idx_t * RP = (rsb_coo_idx_t *) IP;
 
 	const octave_idx_type * OP = (const octave_idx_type*) IP;
 	rsb_nnz_idx_t nzi;
@@ -364,33 +363,33 @@
 	return errval;
 }
 
-static struct rsb_mtx_t *rsboi_mtx_alloc_from_csc_const(const void *VA, /*const*/ octave_idx_type * IA, /*const*/ octave_idx_type * CP, rsb_nnz_idx_t nnzA, rsb_type_t typecode, rsb_coo_idx_t nrA, rsb_coo_idx_t ncA, rsb_blk_idx_t brA, rsb_blk_idx_t bcA, rsb_flags_t flagsA, rsb_err_t *errvalp)
+static struct rsb_mtx_t *rsboi_mtx_alloc_from_csc_const(const void *VA, const octave_idx_type * IA, const octave_idx_type * CP, rsb_nnz_idx_t nnzA, rsb_type_t typecode, rsb_coo_idx_t nrA, rsb_coo_idx_t ncA, rsb_blk_idx_t brA, rsb_blk_idx_t bcA, rsb_flags_t flagsA, rsb_err_t *errvalp)
 {
 	struct rsb_mtx_t *mtxAp = RSBOI_NULL;
 
 	if( RSBOI_SOME_ERROR(rsboi_idx_overflow(errvalp,nrA,ncA,nnzA) ) )
-		goto ret;
-	rsboi_oi2ri(IA,nnzA);
-	rsboi_oi2ri(CP,ncA+1);
-	mtxAp = rsb_mtx_alloc_from_csc_const(VA, (rsb_coo_idx_t *)IA, (rsb_coo_idx_t *)CP, nnzA, typecode, nrA, ncA, brA, bcA, flagsA, errvalp);
-	rsboi_ri2oi((rsb_coo_idx_t *)IA,nnzA);
-	rsboi_ri2oi((rsb_coo_idx_t *)CP,ncA+1);
-ret:
+		return mtxAp;
+	OCTAVE_LOCAL_BUFFER(rsb_coo_idx_t, RIA, nnzA);
+	OCTAVE_LOCAL_BUFFER(rsb_coo_idx_t, RCP, ncA+1);
+	rsboi_oi2ri(IA,RIA,nnzA);
+	rsboi_oi2ri(CP,RCP,ncA+1);
+	mtxAp = rsb_mtx_alloc_from_csc_const(VA, RIA, RCP, nnzA, typecode, nrA, ncA, brA, bcA, flagsA, errvalp);
+
 	return mtxAp;
 }
 
-static struct rsb_mtx_t *rsboi_mtx_alloc_from_coo_const(const void *VA, /*const*/ octave_idx_type * IA, /*const*/ octave_idx_type * JA, rsb_nnz_idx_t nnzA, rsb_type_t typecode, rsb_coo_idx_t nrA, rsb_coo_idx_t ncA, rsb_blk_idx_t brA, rsb_blk_idx_t bcA, rsb_flags_t flagsA, rsb_err_t *errvalp)
+static struct rsb_mtx_t *rsboi_mtx_alloc_from_coo_const(const void *VA, const octave_idx_type * IA, const octave_idx_type * JA, rsb_nnz_idx_t nnzA, rsb_type_t typecode, rsb_coo_idx_t nrA, rsb_coo_idx_t ncA, rsb_blk_idx_t brA, rsb_blk_idx_t bcA, rsb_flags_t flagsA, rsb_err_t *errvalp)
 {
 	struct rsb_mtx_t *mtxAp = RSBOI_NULL;
 
 	if( RSBOI_SOME_ERROR(rsboi_idx_overflow(errvalp,nrA,ncA,nnzA) ) )
-		goto ret;
-	rsboi_oi2ri(IA,nnzA);
-	rsboi_oi2ri(JA,nnzA);
-	mtxAp = rsb_mtx_alloc_from_coo_const(VA, (rsb_coo_idx_t *)IA, (rsb_coo_idx_t *)JA, nnzA, typecode, nrA, ncA, brA, bcA, flagsA, errvalp);
-	rsboi_ri2oi((rsb_coo_idx_t *)IA,nnzA);
-	rsboi_ri2oi((rsb_coo_idx_t *)JA,nnzA);
-ret:
+		return mtxAp;
+	OCTAVE_LOCAL_BUFFER(rsb_coo_idx_t, RIA, nnzA);
+	OCTAVE_LOCAL_BUFFER(rsb_coo_idx_t, RJA, nnzA);
+	rsboi_oi2ri(IA,RIA,nnzA);
+	rsboi_oi2ri(JA,RJA,nnzA);
+	mtxAp = rsb_mtx_alloc_from_coo_const(VA, RIA, RJA, nnzA, typecode, nrA, ncA, brA, bcA, flagsA, errvalp);
+
 	return mtxAp;
 }
 #else /* RSBOI_O64_R32 */
